src/visit/dfsvisit.rs

```diff
@@ -60,26 +69,38 @@ impl<B> Control<B> {
 pub trait ControlFlow {
     fn continuing() -> Self;
     fn should_break(&self) -> bool;
+    fn should_prune(&self) -> bool;
 }
 
 impl ControlFlow for () {
     fn continuing() { }
     #[inline]
     fn should_break(&self) -> bool { false }
+    #[inline]
+    fn should_prune(&self) -> bool { false }
 }
 
 impl<B> ControlFlow for Control<B> {
     fn continuing() -> Self { Control::Continue }
     fn should_break(&self) -> bool {
         if let Control::Break(_) = *self { true } else { false }
     }
+    fn should_prune(&self) -> bool {
+        match *self {
+            Control::Prune => true,
+            Control::Continue | Control::Break(_) => false,
+        }
+    }
 }
 
-impl<E> ControlFlow for Result<(), E> {
-    fn continuing() -> Self { Ok(()) }
+impl<C: ControlFlow, E> ControlFlow for Result<C, E> {
+    fn continuing() -> Self { Ok(C::continuing()) }
     fn should_break(&self) -> bool {
         if let Err(_) = *self { true } else { false }
```

Comment:
We can pass through the "should_break" here like in should_prune. I think that's a good way to combine the power of Result with the control flow thing.